### ISSUE TITLE
Deleting attributes from a schema

### ISSUE THREAD

**Issue by crazyscientist** [SUSE/aimaas#21]
It must be possible to remove attribute definitions from a schema.

To such an action some additional requirements apply:

* The delete action must be subject to review (if feature enabled).
* All values (for this attribute) set on entities of the schema must be deleted.
* These value deletions must be tracked like any other deletion of a value.
* The change table may be extended to provide more context (i.e. that the value was deleted due to the removal of the attribute).

**Comment by commanderprice** [comment]
Should the user who approves removal (or who removes attribute if review is not enabled) have not only have update permissions for this schema, but also update permissions for all entities of this schema?🤔

**Comment by crazyscientist** [comment]
@commanderprice, that is a very good point.

When I wrote this issue I had a use case like this is mind:

* There is a technical admin, who's responsible for the schema layout (i.e. keep schema changes to a minimum, so that clients, which access the API directly, do not break).
* There is a less technical data manager, who's responsible for the entity data.

In a setup with maximal permission separation, it might be desirable to give no entity-permissions to the admin and no schema-permissions to the data manager. In that case I imagined that the admin should still be able to remove the attribute from the schema.

I guess the question now is: Did I put too much trust into this imaginary admin?

**Comment by commanderprice** [comment]
> Did I put too much trust into this imaginary admin?

Well, I guess it's possible to have a technical admin who has permissions to update schema, add/edit attributes, but not to delete them, and a senior technical admin (🤔) who has entities update permission on top of that and can delete attributes as well

Or we can just allow to delete attrs without having permissions for updating entities. For me it just seems to be a kind of a dangerous operation since we are removing attribute values for all entities from database. Although we still track these changes

**Comment by crazyscientist** [comment]
Let's approach the topic from a different angle.

If the app would not automatically delete values in the DB when an admin removes an attribute from a schema, clients would no longer get the values as the API response honors what the schema defines. But the values can be easily recovered, if the attribute gets re-added to the schema again.

At some point we might want to sanitize those inaccessible values:

### Manual solution
* If the review feature is disabled and ...
  * if the admin has the required entity-permission, this can be done right away.
  * if not:
    * The system could remove these values, when an entity gets updated by a data manger.
    * We could over a command to data managers to remove these values.
* If it's enabled, removal of an attribute becomes a bit complicated: We would need one change request for the attribute removal on the schema and one for the value removals on the entities.

Well, this would allow the separation of permissions I had in mind, but complicates the workflow considerably.

### Automatic solution

We could just accept _dead_ values in the DB and log the attribute removal.

As a future enhancement (via #7) we could implement an automatic removal of these values, which is delayed by a configurable interval (e.g. default: 90 days).

This would still maintain separation of permissions and allow quick data recovery.

@commanderprice, what do you think? Does this sound like a less dangerous operation to you?

**Comment by commanderprice** [comment]
@crazyscientist 

> what do you think?

Yeah, I guess we can accept dead values since values are retrieved depending on schema and not the other way around (I hope so 🙄)

I was wondering: when we remove attribute definition from schema, do we delete attribute itself from database if it's not used anywhere else? I guess we shouldn't, since `Value` is linked to a particular `Attribute` and it won't be possible to restore value if attribute doesn't exist anymore

**Comment by crazyscientist** [comment]
> when we remove attribute definition from schema, do we delete attribute itself from database if it's not used anywhere else? I guess we shouldn't.

I agree. IIRC, we might also need the relation to the attribute in the change log/history.

**Comment by commanderprice** [comment]
Oh

We already have a relation to `attr_def` in `ValueUpdate` and `attr` can be retrieved from that relation. But apparently it is not going to work when attribute definition is deleted from schema, so I guess I'll have to rewrite that part🙄. 

Let's suppose we store `attr` instead. Everything works fine until we decide to rename `attr`. As described here https://github.com/SUSE/aimaas/issues/18, renaming would involve

- creating new `attr`
- pointing existing `attr_def` to this `attr`
- pointing all values of old `attr` in current schema to new `attr`
- deleting old `attr` (questionable 🤔)

I guess we need to store relation to `attr` in change history to know which `attr` deleted value belonged to (because for not deleted values we can get this info from `Value` object itself). So, if `attr` was renamed, do we change reference for that `attr` in all `ValueUpdate` entries for that schema? Or there is no need to, because we store the fact of renaming the `attr`?

**Comment by crazyscientist** [comment]
Without looking at the definition of `ValueUpdate` your proposal to replace the foreign key to `attr_def` with a foreign key to `attr`. sounds reasonable :+1: 

From a user's point of view it would probably be more desirable to store an info about the _renaming action_. But we probably would not need to emulate a single _renaming action_ in the history. We could treat it as what it really is: Addition of a new attribute and deletion of another one. This way we would have a consistent change history. :thinking:

**Comment by commanderprice** [comment]
I'm not sure about splitting `attr` renaming on schema into two records in `AttributeCreate` and `AttributeDelete`. Because although internally renaming `attr` involves creation of new `attr` and removal of old one (arguable, I think it's better to keep old attribute), the `attr_def` itself stays the same and traceability for schemas is, from my POV, is about changes in `attr_def`s, not `attr`s

So, for updating `attr_def`s on schema I have such model

```
class AttributeUpdate(Base):
    __tablename__ = 'attr_upd_history'
    id = Column(Integer, primary_key=True)
    change_id = Column(Integer, ForeignKey('changes.id'), nullable=False)
    attr_def_id = Column(Integer, ForeignKey('attr_definitions.id'), nullable=False)
    required = Column(Boolean)
    unique = Column(Boolean)
    list = Column(Boolean)
    key = Column(Boolean)
    description = Column(String(128))
    name = Column(String(128))  # <---- this one is for renaming attr
    bind_to_schema = Column(Integer, ForeignKey('schemas.id'))

    change = relationship('Change')
    attr_def = relationship('AttributeDefinition')
```
I guess I'll have to change relation with `AttributeDefinition` to relation with `Attribute`, but anyway. When user creates a schema update request in which name of attribute is changed, we insert new entry into `AttributeUpdate`. When request is applied, we do all that stuff that is needed to rename `attr` on schema, including updating `attr` references on `Value` objects, but we don't log anything to `ValueUpdate`, because values haven't changed

**Comment by crazyscientist** [comment]
You lost me a little with your proposal. What I think you mean is this:

For the purpose of a consistent data model the `Attribute` model stores the attribute's `name`. Because these attributes can be shared across multiple schemas, we don't want to change/delete existing rows of this table.

This _rename action_ appears as a simple change of the `attribute_id` defined in `AttributeDefinition`, which we will track in `AttributeUpdate`.

The additional replacement of `attribute_id`s in the `ValueXXX` models will not be logged, because the values have not changed.

Yes, that makes sense to me :slightly_smiling_face:

**Comment by commanderprice** [comment]
Yep, you got me right😎

**Comment by commanderprice** [comment]
@crazyscientist 
I was also thinking about such case: user deletes attribute from schema and in same request adds an attribute with same name and type. Do we just update existing `attr_def`  (if it's possible) or delete and create a new one, while also "restoring" values for deleted attribute? There can be a problem with the latter, though, if it was of type `FK`, but bound to different schema

Or we can just forbid such strange manipulations 🤔

**Comment by crazyscientist** [comment]
Very nice catch :+1: 

> user deletes attribute from schema and in same request adds an attribute with same name and type

Putting the `FK` type aside, such a _change_ does not really constitute a change, because before and after the request we have the same `attribute_id`. I'd feel comfortable with forbidding such no-op operations.

> There can be a problem with the latter, though, if it was of type `FK`, but bound to different schema

Absolutely, that is a big problem as clients will end up without the ability to get these related items. 

But I'm afraid this problem is not only limited to deleting and adding an attribute in the same request. Unless the tables holding the values are sanitized between the deletion and addition, we will end up with exactly this problem. :scream: 

Now I get the feeling that this approach leads into a swamp of corner cases. Maybe we should just accept the risk of admins being able to delete values when they remove attributes from a schema :thinking:

**Comment by commanderprice** [comment]
As I understand, you are suggesting to delete values from DB (and log that fact) when attribute is deleted from schema? Which also implies that it's enough to have schema update permissions for that? I guess I'm mentally ready for this, but this doesn't annul results of other discussions here:

- Keep `Attribute` in DB even if no schema is referencing it
- Store `Attribute` instead of `AttributeDefinition` in traceability tables, because `AttributeDefinition` can be removed from DB, but not `Attribute`
- Renaming attribute on schema will result in new entry in `AttributeUpdate`; when this change is applied, `attribute`  reference is changed for corresponding `Value` entries, but nothing is logged to `ValueUpdate`, because value didn't change
- Don't allow removing and adding the same attribute (name-type pair)

**Comment by crazyscientist** [comment]
> As I understand, you are suggesting to delete values from DB (and log that fact) when attribute is deleted from schema? 

Yes, just as I wrote in the initial description :joy: 

> Which also implies that it's enough to have schema update permissions for that?

Yes, I came to believe that this is the most forward approach.

> Keep `Attribute` in DB even if no schema is referencing it. Store `Attribute` instead of `AttributeDefinition` in traceability tables, because `AttributeDefinition` can be removed from DB, but not `Attribute`

Yes, that is my understanding, too.

> Renaming attribute on schema will result in new entry in `AttributeUpdate`; when this change is applied, `attribute` reference is changed for corresponding `Value` entries, but nothing is logged to `ValueUpdate`, because value didn't change.

Yes, makes sense :+1: 

> Don't allow removing and adding the same attribute (name-type pair)

Yes. Also, we have a `UniqueConstraint` for name and type on `Attribute`. So, such an operation would not result in any change in `AttributeDefinition` (identical `attribute_id`)). Given our new understanding, this operation would only result in the deletion of all values related to this attribute.
